File: src/connectors/notion.py
"""
Notion API Connector
Create and manage TODOs in Notion database
"""

import logging
from typing import Dict, List, Any, Optional
from datetime import datetime
from notion_client import Client


class NotionConnector:
    """Wrapper for Notion API"""

    def __init__(self, api_token: str, database_id: str):
        """
        Initialize Notion connector

        Args:
            api_token: Notion integration token
            database_id: Database ID for TODOs
        """
        self.api_token = api_token
        self.database_id = database_id
        self.logger = logging.getLogger("nexus.notion")

        self.client = Client(auth=api_token)

    def create_todo(
        self,
        title: str,
        todo_type: str,
        description: str = "",
        date: Optional[datetime] = None,
        confidence: float = 1.0,
        source: str = ""
    ) -> Optional[Dict[str, Any]]:
        """
        Create a TODO in Notion database

        Args:
            title: TODO title
            todo_type: Type (engagement, demande, deadline)
            description: Description details
            date: Associated date
            confidence: Confidence score (0-1)
            source: Source of the TODO (e.g., "Conversation with X")

        Returns:
            Created page object or None if failed
        """
        try:
            properties = {
                "Titre": {
                    "title": [{"text": {"content": title}}]
                },
                "Type": {
                    "select": {"name": todo_type.capitalize()}
                },
                "Statut": {
                    "select": {"name": "À faire"}
                }
            }

            # Add description if provided
            if description:
                properties["Description"] = {
                    "rich_text": [{"text": {"content": description[:2000]}}]
                }

            # Add date if provided
            if date:
                properties["Date"] = {
                    "date": {"start": date.isoformat()}
                }

            # Add confidence score
            properties["Confiance"] = {
                "number": confidence
            }

            # Add source
            if source:
                properties["Source"] = {
                    "rich_text": [{"text": {"content": source[:1000]}}]
                }

            # Create page
            page = self.client.pages.create(
                parent={"database_id": self.database_id},
                properties=properties
            )

            self.logger.info(f"Created TODO: {title} (type: {todo_type})")
            return page

        except Exception as e:
            self.logger.error(f"Failed to create TODO '{title}': {e}")
            return None

    def create_todos_batch(
        self,
        priorities: Dict[str, List[Dict[str, Any]]]
    ) -> Dict[str, int]:
        """
        Create multiple TODOs from priorities

        Args:
            priorities: Dictionary with keys: engagements, demandes, deadlines

        Returns:
            Dictionary with counts of created TODOs per type
        """
        stats = {
            "engagements": 0,
            "demandes": 0,
            "deadlines": 0,
            "total": 0
        }

        # Create engagements
        for item in priorities.get("engagements", []):
            result = self.create_todo(
                title=item.get("title", "Sans titre"),
                todo_type="engagement",
                description=item.get("description", ""),
                confidence=item.get("confidence", 1.0),
                source=item.get("source", "")
            )
            if result:
                stats["engagements"] += 1
                stats["total"] += 1

        # Create demandes
        for item in priorities.get("demandes", []):
            result = self.create_todo(
                title=item.get("title", "Sans titre"),
                todo_type="demande",
                description=item.get("description", ""),
                confidence=item.get("confidence", 1.0),
                source=item.get("source", "")
            )
            if result:
                stats["demandes"] += 1
                stats["total"] += 1

        # Create deadlines
        for item in priorities.get("deadlines", []):
            result = self.create_todo(
                title=item.get("title", "Sans titre"),
                todo_type="deadline",
                description=item.get("description", ""),
                date=item.get("date"),
                confidence=item.get("confidence", 1.0),
                source=item.get("source", "")
            )
            if result:
                stats["deadlines"] += 1
                stats["total"] += 1

        self.logger.info(f"Created {stats['total']} TODOs in Notion")
        return stats

    def get_database_url(self) -> str:
        """Get Notion database URL"""
        return f"https://notion.so/{self.database_id.replace('-', '')}"

    def is_connected(self) -> bool:
        """Check if connector is configured"""
        return bool(self.api_token and self.database_id)

File: supabase/functions/_shared/clients/notion.ts
/**
 * Notion API Client
 * Manage CRM (Leads, Clients, Projects) in Notion workspace
 */

import { Client } from 'npm:@notionhq/client@2.2.15';
import type { NotionLead, NotionClient, NotionProject } from '../types/index.ts';

export class NotionClient {
  private client: Client;
  private workspaceId: string;
  private leadsDatabase?: string;
  private clientsDatabase?: string;
  private projectsDatabase?: string;

  constructor(
    token: string,
    workspaceId: string,
    databases?: {
      leads?: string;
      clients?: string;
      projects?: string;
    }
  ) {
    this.client = new Client({ auth: token });
    this.workspaceId = workspaceId;
    this.leadsDatabase = databases?.leads;
    this.clientsDatabase = databases?.clients;
    this.projectsDatabase = databases?.projects;
  }

  /**
   * Create a new lead in Notion CRM
   * @param lead - Lead information
   * @returns Created page object
   */
  async createLead(lead: NotionLead): Promise<any> {
    if (!this.leadsDatabase) {
      throw new Error('Leads database ID not configured');
    }

    try {
      console.log(`[Notion] Creating lead: ${lead.name}`);

      const response = await this.client.pages.create({
        parent: { database_id: this.leadsDatabase },
        properties: {
          'Nom': {
            title: [{ text: { content: lead.name } }],
          },
          'Entreprise': lead.company
            ? { rich_text: [{ text: { content: lead.company } }] }
            : undefined,
          'Statut': {
            select: { name: this.formatLeadStatus(lead.status) },
          },
          'Dernière interaction': {
            date: { start: lead.lastInteraction.toISOString() },
          },
          'Prochaine action': lead.nextAction
            ? { rich_text: [{ text: { content: lead.nextAction } }] }
            : undefined,
          'Source': lead.source
            ? { rich_text: [{ text: { content: lead.source } }] }
            : undefined,
        },
        children: lead.notes
          ? [
              {
                object: 'block',
                type: 'paragraph',
                paragraph: {
                  rich_text: [{ text: { content: lead.notes } }],
                },
              },
            ]
          : undefined,
      });

      console.log(`[Notion] Lead created: ${response.id}`);
      return response;

    } catch (error) {
      console.error('[Notion] Failed to create lead:', error);
      throw error;
    }
  }

  /**
   * Create a new client in Notion CRM
   * @param client - Client information
   * @returns Created page object
   */
  async createClient(client: NotionClient): Promise<any> {
    if (!this.clientsDatabase) {
      throw new Error('Clients database ID not configured');
    }

    try {
      console.log(`[Notion] Creating client: ${client.name}`);

      const response = await this.client.pages.create({
        parent: { database_id: this.clientsDatabase },
        properties: {
          'Nom': {
            title: [{ text: { content: client.name } }],
          },
          'Entreprise': {
            rich_text: [{ text: { content: client.company } }],
          },
          'Mandat': client.mandate
            ? { rich_text: [{ text: { content: client.mandate } }] }
            : undefined,
          'Statut': {
            select: { name: this.formatClientStatus(client.status) },
          },
          'Début': client.startDate
            ? { date: { start: client.startDate.toISOString() } }
            : undefined,
          'Fin prévue': client.endDate
            ? { date: { start: client.endDate.toISOString() } }
            : undefined,
          'Prochaine livraison': client.nextDeliverable
            ? { rich_text: [{ text: { content: client.nextDeliverable } }] }
            : undefined,
        },
      });

      console.log(`[Notion] Client created: ${response.id}`);
      return response;

    } catch (error) {
      console.error('[Notion] Failed to create client:', error);
      throw error;
    }
  }

  /**
   * Create a new project in Notion
   * @param project - Project information
   * @returns Created page object
   */
  async createProject(project: NotionProject): Promise<any> {
    if (!this.projectsDatabase) {
      throw new Error('Projects database ID not configured');
    }

    try {
      console.log(`[Notion] Creating project: ${project.title}`);

      const response = await this.client.pages.create({
        parent: { database_id: this.projectsDatabase },
        properties: {
          'Titre': {
            title: [{ text: { content: project.title } }],
          },
          'Client': {
            rich_text: [{ text: { content: project.client } }],
          },
          'Statut': {
            select: { name: this.formatProjectStatus(project.status) },
          },
          'Next Action': project.nextAction
            ? { rich_text: [{ text: { content: project.nextAction } }] }
            : undefined,
          'Deadline': project.deadline
            ? { date: { start: project.deadline.toISOString() } }
            : undefined,
        },
      });

      console.log(`[Notion] Project created: ${response.id}`);
      return response;

    } catch (error) {
      console.error('[Notion] Failed to create project:', error);
      throw error;
    }
  }

  /**
   * Search for existing lead by name
   * @param name - Lead name
   * @returns Page object if found, null otherwise
   */
  async findLeadByName(name: string): Promise<any | null> {
    if (!this.leadsDatabase) {
      return null;
    }

    try {
      const response = await this.client.databases.query({
        database_id: this.leadsDatabase,
        filter: {
          property: 'Nom',
          title: {
            contains: name,
          },
        },
      });

      return response.results.length > 0 ? response.results[0] : null;
    } catch (error) {
      console.error('[Notion] Failed to search lead:', error);
      return null;
    }
  }

  /**
   * Update existing lead
   * @param pageId - Notion page ID
   * @param updates - Fields to update
   */
  async updateLead(
    pageId: string,
    updates: Partial<NotionLead>
  ): Promise<any> {
    try {
      const properties: any = {};

      if (updates.status) {
        properties['Statut'] = {
          select: { name: this.formatLeadStatus(updates.status) },
        };
      }

      if (updates.lastInteraction) {
        properties['Dernière interaction'] = {
          date: { start: updates.lastInteraction.toISOString() },
        };
      }

      if (updates.nextAction) {
        properties['Prochaine action'] = {
          rich_text: [{ text: { content: updates.nextAction } }],
        };
      }

      const response = await this.client.pages.update({
        page_id: pageId,
        properties,
      });

      console.log(`[Notion] Lead updated: ${pageId}`);
      return response;

    } catch (error) {
      console.error('[Notion] Failed to update lead:', error);
      throw error;
    }
  }

  /**
   * Format lead status for Notion select
   */
  private formatLeadStatus(status: string): string {
    const statusMap: Record<string, string> = {
      'cold': '⚪ Froid',
      'warm': '🟡 Tiède',
      'hot': '🔥 Chaud',
      'client': '🟢 Client',
    };
    return statusMap[status] || status;
  }

  /**
   * Format client status for Notion select
   */
  private formatClientStatus(status: string): string {
    const statusMap: Record<string, string> = {
      'active': '🟢 Actif',
      'paused': '🟡 Pause',
      'completed': '✅ Complété',
    };
    return statusMap[status] || status;
  }

  /**
   * Format project status for Notion select
   */
  private formatProjectStatus(status: string): string {
    const statusMap: Record<string, string> = {
      'active': '🔵 Actif',
      'waiting': '🟡 Attente',
      'completed': '✅ Complété',
      'someday': '⚪ Someday',
    };
    return statusMap[status] || status;
  }

  /**
   * Health check - verify API connectivity
   */
  async isConnected(): Promise<boolean> {
    try {
      await this.client.users.me({});
      return true;
    } catch {
      return false;
    }
  }
}
